sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DatetimeFormatterSuite.scala

```diff
@@ -21,11 +21,44 @@ import org.scalatest.Matchers
 
 import org.apache.spark.{SparkFunSuite, SparkUpgradeException}
 import org.apache.spark.sql.catalyst.plans.SQLHelper
+import org.apache.spark.sql.catalyst.util.DateTimeTestUtils.{date, UTC}
 
 trait DatetimeFormatterSuite extends SparkFunSuite with SQLHelper with Matchers {
   import DateTimeFormatterHelper._
   def checkFormatterCreation(pattern: String, isParsing: Boolean): Unit
 
+  private def dateFormatter(pattern: String): DateFormatter = {
+    DateFormatter(pattern, UTC, isParsing = true)
+  }
+
+  private def timestampFormatter(pattern: String): TimestampFormatter =
+    TimestampFormatter(pattern, UTC, isParsing = true)
+
+  protected def useDateFormatter: Boolean
+
+  private def assertEqual(pattern: String, datetimeStr: String, expected: Long): Unit = {
+    if (useDateFormatter) {
+      assert(dateFormatter(pattern).parse(datetimeStr) ===
+        DateTimeUtils.microsToEpochDays(expected, UTC))
+    } else {
+      assert(timestampFormatter(pattern).parse(datetimeStr) === expected)
+    }
+  }
+
+  private def assertError(pattern: String, datetimeStr: String, expectedMsg: String): Unit = {
+    val exception = if (useDateFormatter) {
+      intercept[Exception](dateFormatter(pattern).parse(datetimeStr))
```

Comment:
can we specify the exception class?

Reply:
Ok, for the current test cases, we can capture only `SparkUpgradeException` here.